### DISCUSSION TITLE
RC5x Rendering Question

### DISCUSSION THREAD

**Discussion by elockman** [bengtmartensson/IrScrutinizer#475] [Decoding issues] [answered]
I am trying to render a code for a Marantz SR7015 AVR.
I've had luck with some, but ran into a problem with what I think is an RC5x code.

I've downloaded codes from Marantz, and they document params as (System, Command, Extension).
[IR CODES FOUND HERE](https://www.marantz.com/-/media/files/documentmaster/marantzna/us/marantz_fy21_sr8015_sr7015_ir_code_v01.xls)

Power Off is (16, 12, 2) and works
CD Select is (20, 63, --) and works 
CBL Select is (6, 63, --) and works 
MULTI-CH Stereo is (15,37,57) and works
DOLBY ATMOS is (16,64,25) but throws an exception.

To render, I select _RC5x protocol_, then I enter D=16, S=25, F=64, T=1 and press _Render_, but get the following:
**"DomainViolationException: Parameter F = 64 is outside of the allowed domain: 0..63"**

Am I misundrstanding the Marantz notation? I realize this is the first Command (F param) that is greater than 63. I'm wondering if I need to mask some values and move a bit to make an RC5x command to work with the Marantz notation. 

Does anyone have advice they can offer?

**Comment by bengtmartensson** [comment, marked as answer]
You have to set S=Command and F=Extension (you are doing it the other way around). No, this is not very logical, and yes, it would have been better to have it differently. 

This parametrization was done [many years ago](http://www.hifi-remote.com/wiki/index.php/DecodeIR#RC5x). Unfortunately, fixing it would probably lead to more confusion.

**Reply by elockman** [reply]
@bengtmartensson thank you for the quick response. I had tried that in the past, but it didn't seem to work with other codes.  I'll try that again and report back.

**Reply by elockman** [reply]
Swapping these values do not work with the Marantz parametrization. It also breaks my other IR commands.

If I change the protocol from RC5x to RC5, I can set D=16 and F=64 with no exception, but RC5x limits F from 0-63. I do not know the protocols well enough to understand why.

The Marantz docs show some codes with no Extension, I assumed those were RC5 codes and didn't require and RC5x extension. That is how I came to the conclusion of assigning the parameters.

I am willing to build my own commands manually, but haven't found  a good resource to learn the RC5x protocol. If you can point me to a doc, I'd  appreciate it.

**Reply by elockman** [reply]
I didn't mention this, but I am using the [IRremote](https://github.com/Arduino-IRremote/Arduino-IRremote) repo for IR control in the Arduino IDE.

The issue could be in my IR library. I'll keep looking.

**Reply by elockman** [reply]
@bengtmartensson your original answer was correct. Some of the Marantz codes are RC5 codes, flipping the params broke those decodes. Your original answer is correct.

**Comment by bengtmartensson** [comment]
Right, the commands without "Extension" are RC5, with D=System and F=Command. The other ones are RC5x with D=System, S=Command and F=Extension. (No, this is not a perfect design, admitted.) This answers the question; please mark as "answered".

The protocols (most of them) have a documentation accessible by pressing the "Docu" button on the render pane. Or invoke "Browse protocols" under the "Help" menu. This is all you need.

If you have further problems, please state them in a way that we can help you.

> Swapping these values do not work with the Marantz parametrization. It also breaks my other IR commands.

is not a help-able question.

IrScrutinizer can generate code for IRRemote (an other targets). It has unfortunately not been updated for the current IRRemnote version though #444.
